cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeEngineCredentials.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.runtime.spi.provisioner.dataproc;
+
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.GenericData;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.common.io.CharStreams;
+import com.google.gson.Gson;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+
+/**
+ * Provides ComputeEngineCredentials either locally if no endpoint is defined, or remotely if endpoint is defined.
+ */
+public final class ComputeEngineCredentials extends GoogleCredentials {
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeEngineCredentials.class);
+  private static final Gson GSON = new Gson();
+  /**
+   * Time (in millisecond) to refresh the credentials before it expires.
+   */
+  private static final int NUMBER_OF_RETRIES = 10;
+  private static final int MIN_WAIT_TIME_MILLISECOND = 500;
+  private static final int MAX_WAIT_TIME_MILLISECOND = 10000;
+  private final String endPoint;
+
+  public ComputeEngineCredentials(String endPoint) {
+    this.endPoint = endPoint;
+  }
+
+  private AccessToken refreshAccessTokenLocally() throws IOException {
+    try {
+      GoogleCredentials googleCredentials = com.google.auth.oauth2.ComputeEngineCredentials.create();
+      return googleCredentials.refreshAccessToken();
+    } catch (IOException e) {
+      throw new IOException("Unable to get credentials from the environment. "
+                              + "Please explicitly set the account key.", e);
+    }
+  }
+
+  private AccessToken refreshAccessTokenRemotely(String endPoint) throws IOException {
+    ExponentialBackOff backOff = new ExponentialBackOff.Builder()
+      .setInitialIntervalMillis(MIN_WAIT_TIME_MILLISECOND)
+      .setMaxIntervalMillis(MAX_WAIT_TIME_MILLISECOND).build();
+
+    Exception exception = null;
+    int counter = 1;
+    while (counter <= NUMBER_OF_RETRIES) {
+      try {
+        return fetchAccessToken(endPoint);
```

Comment:
You can still cache the access token in a local volatile field.

Reply:
refreshAccessToken method is called by OAuth2Credentials, and the returned token will be cached there. Note that ComputeEngineCredentials extends  GoogleCredentials which in turn extends OAuth2Credentials.
So it's not needed to cache the token here.

Reply:
We need to cache the token across DataprocConf instances. I introduced a new class called Cached ComputeEngineCredentials thats ComputeEngineCredentials, and provides caching.

Reply:
Update: I rolled the caching mechanism into ComputeEngineCredentials.